Sources/SolToBoogie/ProcedureTranslator.cs

```diff
@@ -1199,37 +1199,49 @@ public override bool Visit(ExpressionStatement node)
                 {
-                    var typeStr = unaryOperation.SubExpression.TypeDescriptions.TypeString;
-                    if (typeStr.StartsWith("int") || typeStr.StartsWith("uint") || typeStr.Equals("bool") || typeStr.StartsWith("string "))
-                    {
-                        //REFACTOR to define isInt, isUint, IsByte, IsString etc.
-                        BoogieExpr rhs = null;
-                        if (typeStr.StartsWith("int") || typeStr.StartsWith("uint"))
-                            rhs = new BoogieLiteralExpr(BigInteger.Zero);
-                        else if (typeStr.Equals("bool"))
-                            rhs = new BoogieLiteralExpr(false);
-                        else if (typeStr.StartsWith("string"))
-                        {
-                            var emptyStr = "";
-                            rhs = new BoogieLiteralExpr(new BigInteger(emptyStr.GetHashCode()));
+                    if (unaryOperation.SubExpression is IndexAccess)
+                    {
+                        var typeStr = unaryOperation.SubExpression.TypeDescriptions.TypeString;
+                        if (typeStr.StartsWith("int") || typeStr.StartsWith("uint") || typeStr.Equals("bool") || typeStr.StartsWith("string "))
+                        {
+                            //REFACTOR to define isInt, isUint, IsByte, IsString etc.
+                            BoogieExpr rhs = null;
+                            if (typeStr.StartsWith("int") || typeStr.StartsWith("uint"))
+                                rhs = new BoogieLiteralExpr(BigInteger.Zero);
+                            else if (typeStr.Equals("bool"))
+                                rhs = new BoogieLiteralExpr(false);
+                            else if (typeStr.StartsWith("string"))
+                            {
+                                var emptyStr = "";
+                                rhs = new BoogieLiteralExpr(new BigInteger(emptyStr.GetHashCode()));
+                            }
+                            var assignCmd = new BoogieAssignCmd(lhs, rhs);
+                            currentStmtList.AddStatement(assignCmd);
+                        }
+                        else
+                        {
+                            Console.WriteLine($"Warning!!: Only handle delete for scalars, found {typeStr}");
                         }
-                        var assignCmd = new BoogieAssignCmd(lhs, rhs);
-                        currentStmtList.AddStatement(assignCmd);
                     }
-                    else
-                    {
-                        Console.WriteLine($"Warning!!: Only handle delete for scalars, found {typeStr}");
-                    }
-                }
-                return false;
-            }
-            else
-            {
-                // distribute to different visit functions
-                return true;
+                    else if (unaryOperation.SubExpression is Identifier)
```

Comment:
How can you assume that this identifier has an array type? what if this is a scalar variable?

Reply:
Actually what is the semantics of {uint x; x = 33; delete x; }

Reply:
You are right.
I add some (maybe rustic) checking for the identifier being of array type. So it is working for `delete a` with `a` being an array.

BTW, I don't understand the semantics for delete a[i], I was expected the length to be decreased. When I tried with remix I got an error complaining about this statement.

Reply:
Here is what I see from Solidity webpage:

https://solidity.readthedocs.io/en/v0.4.21/types.html

delete a assigns the initial value for the type to a. I.e. for integers it is equivalent to a = 0, but it can also be used on arrays, where it assigns a dynamic array of length zero or a static array of the same length with all elements reset. For structs, it assigns a struct with all members reset.

Are you capturing all the cases, including static arrays?

Reply:
@shuvendu-lahiri I add more support to one more construct which appears to be used very frequently (i.e., delete the whole dynamic array)
Unfortunately I do not include support for all cases. You can include this parcial support  in the main branch or wait until all cases are included.

Reply:
I think the change will added a regression, that is not caught by the test. 
    uint a = 5;
    delete a;
    assert(a == 0);
Previously, this assert would have passed, but since you only look for "delete a[i]" and "delete a:array", you will miss this case.

Reply:
Good catch. Let me try to fix this.

Reply:
@shuvendu-lahiri  I fix this issue and refactor the code a little bit. I also identify the case for static arrays but I still don't handle it (I just issue a warning)